library/std/src/io/mod.rs

```diff
@@ -2559,7 +2559,7 @@ impl<T: Read> Read for Take<T> {
 
         let max = cmp::min(buf.len() as u64, self.limit) as usize;
         let n = self.inner.read(&mut buf[..max])?;
-        self.limit -= n as u64;
+        self.limit = self.limit.saturating_sub(n as u64);
```

Comment:
I would prefer if this would panic instead. Panicks are for indicating that a bug in the program has been detected. A `Read` impl that claims to have written more bytes than it could possibly have written is a bug in the program.

Gonna poke @rust-lang/libs to confirm the desired behavior.

Reply:
:+1: on always panicking in that scenario rather than overflowing

Reply:
Oh, I didn't realize that. It helps the user to find the bug if `read` reports wrong size of read contents.

Reply:
How about to add an assertion like `assert!(n as u64 <= self.limit, "number of read bytes exceed ``Take``'s limit");` here?